### PR TITLE
[WIP] Add unit tests to smoke test OpenAPI delegation

### PR DESCRIPTION
**Pull request by frobware**
**What this PR does / why we need it**:

Add a smoke test that checks for the existence of some representative
paths from the apiextensions-server and the kube-aggregator server,
both part of the delegation chain in kube-apiserver.

And although the /api/openapi/swagger.json has the whole spec in it
and it is checked in, it is no guarantee that nothing is broken. This test
is a smoke test to help ensure we don't break OpenAPI delegation.
It's explicit, the checked in json file is not.

**Which issue this PR fixes**: 

Validates at test time #50011

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

File: cmd/kube-apiserver/app/testing/server_test.go [hunk 1]
```diff
@@ -19,13 +19,15 @@ package testing
...
 	"fmt"
+	"strings"
 	"testing"
 	"time"
 
 	admissionregistrationv1alpha1 "k8s.io/api/admissionregistration/v1alpha1"
 	appsv1beta1 "k8s.io/api/apps/v1beta1"
 	corev1 "k8s.io/api/core/v1"
 	networkingv1 "k8s.io/api/networking/v1"
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
```
Comment (sttts) [inline cmd/kube-apiserver/app/testing/server_test.go]: nit: can you reference the v1beta1 package?

File: cmd/kube-apiserver/app/testing/server_test.go [hunk 2]
```diff
@@ -38,6 +40,7 @@ import (
 	utilfeaturetesting "k8s.io/apiserver/pkg/util/feature/testing"
 	"k8s.io/client-go/dynamic"
 	"k8s.io/client-go/kubernetes"
+	"k8s.io/kube-aggregator/pkg/apis/apiregistration"
```
Comment (sttts) [inline cmd/kube-apiserver/app/testing/server_test.go]: same here

**Comment by frobware** [comment]
> nit: can you reference the v1beta1 package?

Done.

**Comment by mbohlool** [comment]
> And although the /api/openapi/swagger.json has the whole spec in it
and it is checked in, it is no guarantee that nothing is broken. This test
is a smoke test to help ensure we don't break OpenAPI delegation.
It's explicit, the checked in json file is not.

As I discussed offline with @sttts, I am against this change because this test guarantee even less than what checked-in spec is guarantees (checked in spec covers whole surface of kubernetes APIs while this one only checks for samples). The argument that reviewers or authors will ignore changes in the checked-in spec is not valid. The spec is checked in mostly to validates PR's change to the surface of the API. If anybody ignores that, they are simply wrong. @bgrant0607 @lavalamp

**Comment by sttts** [comment]
@mbohlool it is not a replacement of the checked in spec. It's just a unit/integration test, like all those 485 other tests we have. Most of them are incomplete. Many are overspecified. Still we have them to ensure that mistakes are avoided. That's the nature of tests. They are subjective because some developer fears regressions without a test. Honestly, I have not seen many discussions about being against a test.

**Comment by bgrant0607** [comment]
@mbohlool Are you opposed to adding a test or to removing a test?

**Comment by frobware** [comment]
Note: currently expecting 'bazel-test' to fail. This PR is dependent on #50864.

**Comment by mbohlool** [comment]
/lgtm 

Please squash commits.

**Comment by mbohlool** [comment]
@bgrant0607 I was concern about the fact that people may ignore checked in spec and we are adding these kind of tests to cover that. I guess we can submit this test but follow up on this after code freeze to add an extra process to checked-in spec so people don't ignore it. Maybe we can add an extra label like "api-change-approved" and "api-change-approval-required" and bot can set "api-change-approval-required" to any PR that touches OpenAPI spec.

**Comment by frobware** [comment]
Note: expecting both `pull-kubernetes-unit` and `pull-kubernetes-bazel-test` to fail until #50864 merges.

**Comment by liggitt** [comment]
TestOpenAPIDelegationChainPlumbing is failing unit test

**Comment by frobware** [comment]
> TestOpenAPIDelegationChainPlumbing is failing unit test

I believe this is failing because because we now exceed the overall test timeout of 5mins. Note, this test does not ordinarily run for 5 minutes. I think all the log spam and the previous tests which are still trying to make connections has tipped this over the timeout threshold. If you run this test independently it completes OK. And, if you move this test to the top of file (i.e., reoprder them) then this test passes and the last test declared then fails.

**Comment by mbohlool** [comment]
is checking race conditions important for this test? One source of long running tests is race condition check that can be disabled by "// +build !race" at the beginning of the test file. If you are not sure, maybe we should move this test to a new file and only add the tag there.

**Comment by frobware** [comment]
> One source of long running tests is race condition check that can be disabled by // +build !race at the beginning of the test file. If you are not sure, maybe we should move this test to a new file and only add the tag there.

That seems worth trying. At the very least we would temporarily unblock the status quo.

**Comment by sttts** [comment]
> by // +build !race at the beginning of the test file

I am fine with adding that for the whole file.

**Comment by frobware** [comment]
@sttts PTAL as the last test run succeeded now that we use // +build !race.

**Comment by liggitt** [comment]
>> by // +build !race at the beginning of the test file
> 
> I am fine with adding that for the whole file.

Big -1, this just skips the tests when unit tests are run with race detection. This test didn't even run in the "successful" pull-kubernetes-unit job (and adding the build flag meant the other tests in that file stopped running as well)

**Comment by frobware** [comment]
> Big -1, this just skips the tests when unit tests are run with race detection. This test didn't even run in the "successful" pull-kubernetes-unit job (and adding the build flag meant the other tests in that file stopped running as well)

I was just verifying the logs and noticed that this devolved to "no tests found". :(

**Comment by mbohlool** [comment]
> Big -1, this just skips the tests when unit tests are run with race detection (which is how we run tests). This test didn't even run in the "successful" pull-kubernetes-unit job (and adding the build flag meant the other tests in that file stopped running as well)

Do we run tests with and without race detection both? I think moving test to a new file and disable race detection as I suggested should work in any case.

**Comment by frobware** [comment]
This test will not pass CI at the moment because, as a unit test in `cmd/kube-apiserver/app/testing`, it crosses a timeout threshold  (currently 5 mins). This is due to a number of issues:

- CI uses `-race` (exacerbating the time to run)
- we don't close resources down (Storage TCP/client connections), resulting in leaked file descriptors
- massive amount of logspam from lots of go routines (Issue #49489), again exacerbated by `-race`, slowing test execution down

To mitigate the `-race` CI issues I moved this unit test to become an integration test in `test/integration/master` and rebased the changes on top of #51685.

However, this still doesn't pass CI because the existing integration tests (i.e., `master_test.go`) and now `server_test.go` cause the overall test executable to fail with `EMFILE (too many open files)`. We're leaking file descriptors when closing connections to the server. This is purely a limit issue - the combination of the two _test.go files means that the test process has >= 1024 file descriptors open, and any remaining tests in server_test.go fail. If you temporarily disable the tests in `master_test.go` then it passes. And vice-versa.

I experimented by rebasing this PR against #50690 and that indeed works up to a point. I say "a point" because although #50690 does help close open file descriptors when the `teardown()` function is called, there is still significant leakage. As an experiment (with #50690 included) you can run 48 instances of just the following  before you'll run out of file descriptors again:

````
func ttt(t *testing.T) {
	_, _, tearDown := kubeapiserver.StartTestServerOrDie(t, nil, framework.SharedEtcd())
	defer tearDown()
}

func TestFoo(t *testing.T) {
	for i := 0; i < 100; i++ {
		ttt(t)
	}
}

````
It seems we have a number of options (and canvassing for other solutions too):

1. we question the usefulness of this test, ignore all problems and move on (gag!)
2. we take another look at the criticisms raised against #50690 
3. we raise thresholds in CI, temporarily walking away from the issues found here 
4. we investigate @smarterclayton "integration runner"
5. we continue to hunt down FD leaks
6. we put each TestXXX function in its own package/test executable (gag!)
7. _and I'm sure there are others_

Right now this is consuming significant amount of test/CI/debug effort and I wanted to stop to see what makes most sense.

/cc @sttts @ironcladlou @derekwaynecarr @mbohlool

**Comment by dims** [comment]
@frobware please rebase

**Comment by mbohlool** [comment]
@frobware did you try to keep the test as a unit test but in its own file and disable `-race` for that file only? what happened in that case?

**Comment by frobware** [comment]
> @frobware did you try to keep the test as a unit test but in its own file and disable -race for that file only? what happened in that case?

AIUI, the test in its own file with `-race` would not be run in via `make test`.

It's not 100% clear to me whether `make bazel-test` is run without `-race`. If it is, then that would be good as it will get some testing.

**Comment by dims** [comment]
Looks like this won't make it to 1.8. Do we need this?

**Comment by frobware** [comment]
> Looks like this won't make it to 1.8. Do we need this?

Without fixing the issues highlighted w.r.t CI this cannot land.

Admittedly this is a smoke test only and can be skipped for 1.8.

**Comment by dims** [comment]
@frobware drat! looks like we need a rebase now :(

**Comment by frobware** [comment]
> @frobware drat! looks like we need a rebase now :(

Will rebase.

**Comment by frobware** [comment]
> Will rebase.

And I just remembered that this PR depends on #51685, so I abandoned my rebase until we decide how this should land. /cc @sttts

**Comment by frobware** [comment]
I went back to having the unit tests run in `k8s.io/kubernetes/cmd/kube-apiserver/app/testing` but split `TestOpenAPIDelegationChainPlumbing()` into its own file that also has `// +build !race` set. The intent here is to see if the test will run via `make bazel-test`.

**Comment by frobware** [comment]
@sttts shall I just make this WIP. There's no point running the tests as we [used to] know that they will timeout.

**Comment by sttts** [comment]
@frobware yes, please. But leave it open. This topic is still on the agenda.

**Comment by sttts** [comment]
We have that with #56328

**Pull request closed without merging**